lib/graph/helpers/ui5Framework.js

```diff
@@ -4,58 +4,112 @@ import {getLogger} from "@ui5/logger";
 				if (this._libraryMetadata[depName]) {
 					log.verbose(`Resolving optional dependency ${depName} for project ${libName}...`);
-					await this.addProjectToGraph(depName, projectGraph);
+					await this.addProjectToGraph(depName, [...ancestors, libName]);
 					return depName;
 				}
 			}));
 
 			dependencies.push(...resolvedOptionals.filter(($)=>$));
 		}
 
-		const ui5Module = new Module({
-			id: depMetadata.id,
-			version: depMetadata.version,
-			modulePath: depMetadata.path
-		});
+		let projectIsFromWorkspace = false;
+		let ui5Module;
+		if (this._workspace) {
+			ui5Module = await this._workspace.getModuleByProjectName(libName);
+			if (ui5Module) {
+				log.info(`Resolved project ${libName} via ${this._workspace.getName()} workspace ` +
+					`to version ${ui5Module.getVersion()}`);
+				log.verbose(`  Resolved module ${libName} to path ${ui5Module.getPath()}`);
+				log.verbose(`  Requested version was: ${depMetadata.version}`);
+				projectIsFromWorkspace = true;
+			}
+		}
+
+		if (!ui5Module) {
+			ui5Module = new Module({
+				id: depMetadata.id,
+				version: depMetadata.version,
+				modulePath: depMetadata.path
+			});
+		}
 		const {project} = await ui5Module.getSpecifications();
-		projectGraph.addProject(project);
+		graph.addProject(project);
 		dependencies.forEach((dependency) => {
-			projectGraph.declareDependency(libName, dependency);
+			graph.declareDependency(libName, dependency);
 		});
+		if (projectIsFromWorkspace) {
+			// Add any dependencies that are only declared in the workspace resolved project
+			// Do not remove superfluous dependencies (might be added later though)
+			await Promise.all(project.getFrameworkDependencies().map(async ({name, optional, development}) => {
+				// Only proceed with dependencies which are not "optional" or "development",
+				// and not already listed in the dependencies of the original node
+				if (optional || development || dependencies.includes(name)) {
+					return;
+				}
+
+				if (!this._libraryMetadata[name]) {
+					throw new Error(
+						`Unable to find dependency ${name}, required by project ${project.getName()} ` +
+						`(resolved via ${this._workspace.getName()} workspace) in current set of libraries. ` +
+						`Try adding it temporarily to the root project's dependencies`);
+				}
+
+				// TODO: If a cyclic dependency is declared, this will empty the event loop.
+				// I guess this is a general issue and not limited to projects resolved via workspaces
+				await this.addProjectToGraph(name, [...ancestors, libName]);
+				graph.declareDependency(libName, name);
+			}));
+		}
+	}
+	_checkCycle(ancestors, projectName) {
+		if (ancestors.includes(projectName)) {
+			// "Back-edge" detected. This would cause a deadlock
+			// Mark first and last occurrence in chain with an asterisk and throw an error detailing the
+			// problematic dependency chain
+			ancestors[ancestors.indexOf(projectName)] = `*${projectName}*`;
```

Comment:
same as above. Why not marking all occurrences?

Reply:
See above

Reply:
ack.